core/src/main/java/org/jdbi/v3/core/mapper/reflect/InstanceFactory.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.mapper.reflect;
+
+import javax.annotation.Nullable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Parameter;
+
+import static java.util.Objects.requireNonNull;
+
+abstract class InstanceFactory<T> {
+    private final Executable executable;
+
+    protected InstanceFactory(Executable executable) {
+        this.executable = requireNonNull(executable, "executable is null");
+    }
+
+    public Class<?> getDeclaringClass() {
+        return executable.getDeclaringClass();
+    }
+
+    public int getParameterCount() {
+        return executable.getParameterCount();
+    }
+
+    public Parameter[] getParameters() {
+        return executable.getParameters();
+    }
+
+    @Nullable
+    public <A extends Annotation> A getAnnotation(Class<A> annotationClass) {
+        return executable.getAnnotation(annotationClass);
+    }
+
+    public abstract T newInstance(Object... params);
```

Comment:
Nit: since class is package private, all the above methods should be package private too

Reply:
as per https://github.com/jdbi/jdbi/pull/1316#discussion_r236820179, i made this class public

Reply:
Ah, I didn't think this one through--didn't mean to expose `InstanceFactory` as API. In my head I was thinking it would return an `Executable` but that's probably pointless since it has to be cast to `Constructor` or `Method` to invoke it.

Let's make this class package private again, and hide that static method in `JdbiConstructors like it was before. Sorry for the churn.